<a href="https://colab.research.google.com/github/Vitaly-lv/DS2022/blob/main/NLP1_homework_%D0%9B%D0%B5%D0%BE%D0%BD%D1%82%D1%8C%D0%B5%D0%B2%D0%92%D0%92.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

###Задание 5: Подсчитайте количество разных слов



In [ ]:
#Установка нужных пакетов
!pip install --upgrade nltk gensim bokeh umap-learn

import itertools
import string

import numpy as np
import umap
from nltk.tokenize import WordPunctTokenizer

from matplotlib import pyplot as plt

from IPython.display import clear_output

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 24.1 MB 1.5 MB/s 
     |████████████████████████████████| 16.4 MB 133 kB/s 
     |████████████████████████████████| 88 kB 2.2 MB/s 
     |████████████████████████████████| 55 kB 1.8 MB/s 
     |████████████████████████████████| 295 kB 4.5 MB/s 
     |████████████████████████████████| 1.1 MB 52.3 MB/s 
  Created wheel for umap-learn: filename=umap_learn-0.5.3-py3-none-any.whl size=82829 sha256=dbf9bc2ac0662893022ba52aa0b4f74c93ec0994b925f0174839f53025b553a3
  Stored in directory: /root/.cache/pip/wheels/a9/3a/67/06a8950e053725912e6a8c42c4a3a241410f6487b8402542ea
  Created wheel for pynndescent: filename=pynndescent-0.5.8-py3-none-any.whl size=55513 sha256=17738d30c074fc074676e33a1a7d8c6222745f0bf7a325cb3af1eec9c47400f3
  Stored in directory: /root/.cache/pip/wheels/1c/63/3a/29954bca1a27ba100ed8c27973a78cb71b43dc67aed62e80c3
Successfully built umap-le

In [ ]:
# выгружаем датасет:
!wget https://www.dropbox.com/s/obaitrix9jyu84r/quora.txt?dl=1 -O ./quora.txt -nc

--2022-12-07 06:33:31--  https://www.dropbox.com/s/obaitrix9jyu84r/quora.txt?dl=1
Resolving www.dropbox.com (www.dropbox.com)... 162.125.2.18, 2620:100:6017:18::a27d:212
Connecting to www.dropbox.com (www.dropbox.com)|162.125.2.18|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: /s/dl/obaitrix9jyu84r/quora.txt [following]
--2022-12-07 06:33:31--  https://www.dropbox.com/s/dl/obaitrix9jyu84r/quora.txt
Reusing existing connection to www.dropbox.com:443.
HTTP request sent, awaiting response... 302 Found
Location: https://uc698577a49b143f0e9f23ace7ed.dl.dropboxusercontent.com/cd/0/get/ByIjIhW7F8r_FPykG_NlLwT9iEE6T4fA4Wx56ODzSwh8tzOpmumI3MkdE7VXss53apDaT1Ii1MNvGRgWEvTDFrcRULKy2gAhOpqiPeVwg4DlAfEEkUV1m80TE12-lY2xc1ZuB7CYbK5ki2UWhrz50mnvSN_wxkOg8y3mQ63V9wdODw/file?dl=1# [following]
--2022-12-07 06:33:32--  https://uc698577a49b143f0e9f23ace7ed.dl.dropboxusercontent.com/cd/0/get/ByIjIhW7F8r_FPykG_NlLwT9iEE6T4fA4Wx56ODzSwh8tzOpmumI3MkdE7VXss53apDaT1Ii1MNvGRgWEvTDFrc

In [ ]:
data = list(open("./quora.txt", encoding="utf-8"))
data[50]

"What TV shows or books help you read people's body language?\n"

In [ ]:
# Токенизация и приведение к нижнему регистру
tokenizer = WordPunctTokenizer()
data_tok = []
for x in data:
    data_tok.append(tokenizer.tokenize(x.lower()))

In [ ]:
import nltk
nltk.download('wordnet')
nltk.download('omw-1.4')
from nltk.stem import WordNetLemmatizer
from nltk.stem import PorterStemmer
from nltk.tokenize import word_tokenize

ps = PorterStemmer()
lemmatizer = WordNetLemmatizer()

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...


1.Подсчет количество разных слов сначала лемматизация потом стэмминг

In [ ]:
# Лемматизация
all_words = []
for i in data_tok:
  for j in i:
    all_words.append(j)

unique_words = set(all_words)
all_lem = [lemmatizer.lemmatize(i) for i in unique_words] #Лемматизация только уникальных слов
# Проверим есть ли повторяющиеся лемматизированные слова
print(len(all_lem),len(set(all_lem))) 

87819 80303


In [ ]:
unique_all_lem = set(all_lem) #так как есть повторяющиеся лемматизированные слова для стэмминга возьмем только уникальные

In [ ]:
# Стэмминг
all_st = [ps.stem(i) for i in unique_all_lem]

print(len(all_st),len(set(all_st)))

80303 66835


2.Подсчет количество разных слов сначала стэмминг потом лемматизация  

In [ ]:
# Стэмминг
all_words = []
for i in data_tok:
  for j in i:
    all_words.append(j)

unique_words = set(all_words)
# аналогично берем только уникальные слова
all_st = [ps.stem(i) for i in unique_words]

print(len(all_st),len(set(all_st)))

87819 67026


In [ ]:
unique_all_st = set(all_st) #для лемматизации возьмем только уникальные слова после стэмминга

In [ ]:
all_lem = [lemmatizer.lemmatize(i) for i in unique_all_st]

print(len(all_lem),len(set(all_lem)))

67026 66818


В первом случае (сначала лемматизация потом стемминг) получилось 66835 разных слов, во втором случае (сначала стэмминг потом лемматизация) получилось 66818 разных слов.

###ML1_1: 
https://www.hackerrank.com/challenges/capturing-non-capturing-groups/problem?isFullScreen=true


(?:[o][k]){3,}

###ML1_2: 
https://www.hackerrank.com/challenges/branch-reset-groups/problem?isFullScreen=true

'/(?|(\d\d-\d\d-\d\d-\d\d$)|(^\d\d---\d\d---\d\d---\d\d)|(\d\d\.\d\d\.\d\d\.\d\d)|(\d\d:\d\d:\d\d:\d\d))/'
